Building Trove classifier tree using markup

waseem18 opened the pull request:
- Fixes #1299 
- Changed Indentation based filter sidebar to markup based along with styling.

waseem18 commented:
Result: 

![11](https://user-images.githubusercontent.com/5843007/37489847-ed62d3c4-28be-11e8-9d18-7aa160def8d1.png)

![33](https://user-images.githubusercontent.com/5843007/37489841-e843c54c-28be-11e8-85eb-8eee912bf8ca.png)

waseem18 commented:
@nlhkabu In this WIP PR I tried changing indentation based Trove classifier filter to markup based. Can you please review this PR?

Thanks

nlhkabu commented:
sure can @waseem18 !  I'll try to take a look over the weekend!  Thanks for the start on this :)

waseem18 commented:
I see lint errors in `_checkbox-tree.scss` file - I'll fix them.

brainwane commented:
Looks like this is passing checks now. Thanks, @waseem18! Looks like this is ready for re-review.

waseem18 commented:
Yes @brainwane

lgh2 commented:
Hi @waseem18 ! I took a look at this with Chrome Version 65.0.3325.124 (Official Build) beta (64-bit), Firefox 58.0.2 (64-bit), and Safari on my iPhone.  It looks really nice, and I didn't find any issues.

brainwane commented:
@yeraydiazdiaz I'd appreciate it if you could review this.

yeraydiazdiaz commented on warehouse/templates/search/results.html:
```diff
@@ -84,6 +84,32 @@ <h3 class="package-snippet__title">
+    {%- for classifier in classifiers %}
+     {% set children = classifiers[classifier] %}
+     {% if kwargs['parent'] != None %}
+      {% set _id = _id +' :: '+ kwargs['parent'] %}
```
could you rename `_id` to something like `level_id`? it would be clearer what the purpose of it is in the line below

yeraydiazdiaz commented on warehouse/templates/search/results.html:
```diff
@@ -84,6 +84,32 @@ <h3 class="package-snippet__title">
+    {% set _id = top_level %}
+    {%- for classifier in classifiers %}
+     {% set children = classifiers[classifier] %}
+     {% if kwargs['parent'] != None %}
```
it's good practice to use `is not None` instead of `!= None`

yeraydiazdiaz commented on warehouse/templates/search/results.html:
```diff
@@ -84,6 +84,32 @@ <h3 class="package-snippet__title">
+    {% set id = _id + ' :: ' + classifier %}
+    <div class="checkbox-tree__item">
+      <li>
+        <input name="c" type="checkbox" id="{{ id }}" class="-js-form-submit-trigger checkbox-tree__checkbox" value="{{ id }}" {{ 'checked' if id in applied_filters else '' }}>
```
is `applied_filters` available in the scope of this macro? you may need to pass it as an extra argument

waseem18 replied:
@yeraydiazdiaz `applied_filters` is available in the scope `generate_markup` macro. We wouldn't need to pass it as an argument.

yeraydiazdiaz replied:
I understand, but I think it might be more explicit to take it as an argument. For example if we were to move the macro to a different file a developer reading the code would wonder what the value for `applied_filters` would be.

yeraydiazdiaz commented on warehouse/static/sass/blocks/_checkbox-tree.scss:
```diff
@@ -39,3 +43,41 @@
+    margin: 0;
+    padding: 3px 7px;
+    line-height: 20px;
+    color: #464646;
```
this particular color is part of the [predefined colors](https://github.com/pypa/warehouse/blob/master/warehouse/static/sass/settings/_colours.scss#L30) you should be able to just use `color: $text-color;` here

nlhkabu commented:
Hi @waseem18 - thanks for this PR! Thanks also to @yeraydiazdiaz for your review :)

Right now, we have a `<div>` as a nested child of a `<ul>`, which is not valid HTML.  It *looks* as though this div is not doing anything though, so we can remove it?

I don't know if you've already seen this - but this JS fiddle is a good piece of inspiration for the next steps:

https://jsfiddle.net/1fynun7a/1/

waseem18 commented on warehouse/templates/search/results.html:
```diff
@@ -84,6 +84,32 @@ <h3 class="package-snippet__title">
 
+
+
+{% macro generate_markup(top_level, classifiers) -%}
```
@yeraydiazdiaz This macro executes recursively. In [line 104](https://github.com/pypa/warehouse/pull/3273/files/5939af2b844b66180ba85369454d6dcc89b51724#diff-f58bde7114ccc3bb3f834162bbdd560dR104) we've passed the `parent` parameter with the value `classifier` so if we specify `parent=None` is macro arguments it'll override `parent=classifier`.

yeraydiazdiaz replied:
Macros in Jinja2 are essentially Python functions, adding `parent=None` specifies the default value that `parent` should take, so it only assigns `None` to `parent` if the caller does not specify a value for it. Passing `parent=classifier` should be perfectly safe.

waseem18 replied:
Thanks for the response @yeraydiazdiaz 

Will address the review today.

waseem18 replied:
@yeraydiazdiaz Addressed :+1:

waseem18 commented:
Addressed reviews @nlhkabu @yeraydiazdiaz

brainwane commented on warehouse/views.py:
```diff
@@ -326,11 +326,41 @@ def filter_key(item):
+        Takes a list of lists each of them containing filter and
+        one of it's child.
+        Returns dictionary with key being filter and value being
+        it's children
```
Better grammar (assuming this is accurate):

Takes a list of lists, each of them containing a filter and one of its children.
Returns a dictionary, each key being a filter and each value being the filter's children.

brainwane commented on warehouse/views.py:
```diff
@@ -326,11 +326,41 @@ def filter_key(item):
+        """
+        Processes available filters and returns a list of dictionaries
+        Value of a key in the dictionary represents it's children
+        """
```
Better grammar (assuming that my understanding is accurate):

Processes available filters and returns a list of dictionaries.
The value of a key in the dictionary represents its children.

nedbat commented:
I love the hierarchical presentation.  Would this also apply to a single package's page where the classifiers are listed? For example, https://pypi.org/project/coverage/

waseem18 commented:
@nedbat This PR applies the hierarchical presentation only on the search page. However it would be awesome to apply similar presentation on a package's page.

@nlhkabu Your thoughts?

waseem18 commented:
Updated screenshots: 
![1](https://user-images.githubusercontent.com/5843007/38147904-e0e1d8ca-3471-11e8-93d6-0426fdab6f49.png)
![2](https://user-images.githubusercontent.com/5843007/38147905-e2fa4b7e-3471-11e8-804d-20b4c817c3f8.png)

lgh2 commented:
Hi @waseem18 ! I took another look at this, and it looks good on Safari on my iPhone.  I was able to select and deselect the classifier checkboxes, and I didn't have trouble doing it. 

I also checked it in the Chrome mobile device simulator, and it worked there, too.

waseem18 commented:
Thank You for taking time and testing this @lgh2 :+1:

nlhkabu commented:
@waseem18 sorry for the delay on this - it is on my list of things to pull down and review this week. Thanks for your patience.

nlhkabu commented:
hi @waseem, thanks for updating the HTML - it's much better :)

A few things:

1. would it be possible to include horizontal lines?  See https://jsfiddle.net/1fynun7a/1/ for inspiration
2. could we align the lines so that they align with the center of the checkboxes?
3. could we make the lines 2px wide and a slightly lighter shade of grey (see `https://github.com/pypa/warehouse/blob/master/warehouse/static/sass/settings/_colours.scss` for colors)

I agree with @nedbat that it would be nice to add this to the project pages too - let's get it working on the search page first, then make a separate PR for this?

Let me know if you need help with the CSS for any of the above, I'm happy to make a contributing commit :)

waseem18 commented:
Thanks for the review @nlhkabu 


> would it be possible to include horizontal lines? See https://jsfiddle.net/1fynun7a/1/ for inspiration

I'll start experimenting with the horizontal lines.

> could we align the lines so that they align with the center of the checkboxes?

We can surely align them to center of checkboxes.

> could we make the lines 2px wide and a slightly lighter shade of grey (see https://github.com/pypa/warehouse/blob/master/warehouse/static/sass/settings/_colours.scss for colors)

Sure. Will do.

As you said let's get this working on the search page first and then I'll put a PR for having the same on the projects page.

I'll start working on the reviews now. In case I get stuck with something I'll seek your help.

Thanks

waseem18 commented:
@nlhkabu I've addressed your reviews and here are some updated screenshots.

On Google Chrome
![1](https://user-images.githubusercontent.com/5843007/39150862-3b14ecb8-4761-11e8-8563-91a796551c45.png)

On Mozilla Firefox
![2m](https://user-images.githubusercontent.com/5843007/39150927-6d79d0ec-4761-11e8-8f25-1d89868e1635.png)

Mobile: (iPhone 6/7/8)
![m](https://user-images.githubusercontent.com/5843007/39151539-1ba24f4a-4763-11e8-8634-b00248b23f54.png)

nlhkabu commented:
Thanks @waseem18 - looking good!

Two more things:

Would it be possible to update the CSS so that the horizontal line in horizontally centered on the checkbox, and there is even spacing around the checkbox? e.g.

![screenshot from 2018-04-24 08-09-10](https://user-images.githubusercontent.com/3323703/39207322-8fe92cd6-47f7-11e8-9bf1-44e7bc5ce5aa.png)

It would also be great if we could make the line color white (or maybe `$primary-light-color?`) when the classifier list is against the blue background.

Again, let me know if you want/need a hand with this! Thanks again!

brainwane commented:
@waseem18 How's this going?

waseem18 commented:
@brainwane I made some progress on this and I'll push the commit today after I get back home from work.

waseem18 commented:
Before I push in the changes I wanted to put up the updated screenshots so any review coming from them can be resolved.

1.With border color as `$primary-color-light`
![mob](https://user-images.githubusercontent.com/5843007/39597337-ac28dddc-4f32-11e8-916a-c628369e5979.png)

2. With border color as `$white` - I feel `$white` border looks good than `$primary-color-light`
![mob-w](https://user-images.githubusercontent.com/5843007/39597342-af3990a2-4f32-11e8-8ad8-1c5300e1e01e.png)

3. Even spacing around checkbox.
![spacing](https://user-images.githubusercontent.com/5843007/39597512-404e45ec-4f33-11e8-9fd7-177ecc2153e3.png)

malikoth commented:
Looks much better than what is there.  But for what it's worth, the checkboxes and lines are not at all centered, neither vertically nor horizontally, for me.  This is taken on Chrome Version 66.0.3359.139 (Official Build) (64-bit).

<img width="158" alt="screen shot 2018-05-14 at 10 25 59" src="https://user-images.githubusercontent.com/2665937/40010207-515e202e-5761-11e8-8675-d21691e58982.png">

waseem18 commented:
Thanks for providing information about browser. I'll take a look at it.

brainwane commented:
Hi @waseem18 and thanks for your work! How is this going?

waseem18 commented:
Hey @brainwane - Was looking into another PR. Will hop on to this today.

waseem18 commented:
@malikoth Could you please provide me some information about what OS you use? I've tested it on Ubunut Chrome 66, and it seems to work fine. I'm not sure about Chrome on Windows though. Have to check.

malikoth commented:
That was taken on MacOS High Sierra (10.13.4) on a 2016 MBP.

waseem18 commented:
Thanks for the response @malikoth I'll take a look at it on Mac.

waseem18 commented:
Update : I'll push a commit over the weekend which would resolve the issue in Chrome on Mac.

waseem18 commented:
Added new commit which (hopefully) will resolve issue on Chrome - Mac.

Below are screenshots from Safari and Chrome on Mac. For other OS's the screens can be seen as above.

On Safari : 
<img width="273" alt="safari" src="https://user-images.githubusercontent.com/5843007/40579590-b802c50a-6148-11e8-9ea5-b12ab85a57d1.png">

Chrome on Mac : 
<img width="372" alt="chrome_mac" src="https://user-images.githubusercontent.com/5843007/40579595-f8c28008-6148-11e8-88fd-711b9a939708.png">

cc: @malikoth

asmacdo commented:
@waseem18 re:your question on pypa-dev

A couple days ago, black was added to the linter. https://github.com/pypa/warehouse/pull/3367

If you do a rebase, `make lint` should catch the problems, and `make reformat` should fix them.

waseem18 commented:
Thanks for the information @asmacdo

waseem18 commented:
@nlhkabu Could you please take a look at this whenever you're free?

I understand that you must be busy.

Thanks

nlhkabu commented:
thank you @waseem18 !
